### ISSUE TITLE
Use with nixpkgs rustc?

### ISSUE THREAD

**Issue by Gaelan** [cargo2nix/cargo2nix#166]
The examples here all use `nixpkgs-mozilla`. I'd prefer to use the rustc provided by Nixpkgs, if possible—can that be done? If not, what needs to happen for it to work?

This would presumably be a requirement for https://github.com/cargo2nix/cargo2nix/issues/42.

**Comment by psionic-k** [comment]
It can be done.   The moz overlay is not providing much special sauce on top of simply bootstrapping a rust toolchain with `rustc`, `cargo`, and `rust-src` (provided purely for LSP integrations)

This issue has good alignment with the goal of providing a default freeze of the `cargo2nix` overlay a rust toolchain as part of the `Cargo2nix` so that the user no longer has to bootstrap with overlays, which is both boilerplate and a potential rabbit hole of confusion for nix beginners

**Comment by psionic-k** [comment]
I scanned the definitions in nixpkgs but could not conclude that it provides nightly or is compatible with the arbitrary channels that nixpkgs-mozilla makes convenient.  It's unclear that nixpkgs rust is better.

Simplifying the overlay boilerplate will alleviate some of the pain anyway?

**Comment by psionic-k** [comment]
I've added an explicit `rustToolchain` argument that overrides the use of the rust-overlay for advanced users.

https://github.com/cargo2nix/cargo2nix/commit/90e3f29717bf2074544dec02f854e9ad2f26fc38

Merges in #224

**Comment by superboum** [comment]
Just a note to say that we had the same problem. We need to use the rust toolchain provided by nixpkgs because we had compilation issues with Mozilla's one. Indeed, we had an [ABI issue when musl 1.2.0 was released](https://musl.libc.org/time64.html) on Nix. Mozilla's rustc was still built with musl < 1.2.0 and nix shipped musl >= 1.2.0 which lead to compilation breakage. Now, we want a Rust release that is bound to our Nix repository to avoid these problems.  Because Cargo2nix does not support the Rust environment shipped by NixOS, we emulate the structure of the Rust object created by rustOverlay.    In practise, rustOverlay ships rustc+cargo in a single derivation while NixOS ships them in separate ones. We reunite them with symlinkJoin. And we need to use the low level makePackageSet function of cargo2nix instead of the high level one.

So our `default.nix` file looks like that in the end:

```nix
  rustChannel = pkgs.symlinkJoin {
    name ="rust-channel";
    paths = [ 
      pkgs.rustPlatform.rust.rustc
      pkgs.rustPlatform.rust.cargo
    ];
  };

  rustPkgs = pkgs.rustBuilder.makePackageSet {
    inherit packageFun rustChannel release codegenOpts;
    packageOverrides = overrides;
    target = null;

    buildRustPackages = pkgs.buildPackages.rustBuilder.makePackageSet {
      inherit rustChannel packageFun codegenOpts;
      packageOverrides = overrides;
      target = null;
    };
  };
```

**Comment by psionic-k** [comment]
@superboum good example of why to support this use case.  There's more components needed for getting a full dev shell going, but I think I can make something work similar to what you have here and then expose some kind of shortcut.

The recent versions in master or PR branch use oxalica overlay and provide ways to override the entire toolchain.
